### ISSUE TITLE
Is it possible to not block atom while goimports is running?

### ISSUE THREAD

**Issue by dominicbarnes** [joefitzgerald/gofmt#47]
This is partly a question, but I'm also partly raising a UX issue. I've been using `goimports` as it can be _really_ convenient. However, if I save a file that happens to have a weird identifier or something it causes `goimports` to scan _everything_ and it's pegging the CPU for long periods of time on end. (it's still running as I write this post)

I guess my main question is... is it necessary to block Atom while this is running? Can't it be run in the background, and cancelled if another save is triggered? Or at the very least, can some sort of timeout be configured so I don't have to wait forever?

**Comment by lloiser** [comment]
First of all make sure to update the responsible tool `golang.org/x/tools/cmd/goimports` either by using the `go get -u golang.org/x/tools/cmd/goimports` or the `Golang: Update Tools` command in atom.

Lets assume `goimports` runs in the background and runs for several seconds (to minutes in your case). `goimports` actually takes the file name as input and writes the formatted code back to disk. So you if change your code while `goimports` is running in the background it will most likely overwrite your changes.

Allowing it to be canceled and the timeout seems like reasonable features though!

**Comment by arxdsilva** [comment]
@lloiser I had a similar issue, when using in small projects `goimports` would work perfectly, but when I tried to use in bigger projects It would not only crash Atom but start an infinite `goimports` loop loading up the CPU, even with the latest version of `goimports`. 

The only way to make the package work with my project was to use `gofmt` as the format tool. Maybe using this FormatTool as default would prevent this problem on new users.

**Comment by lloiser** [comment]
Can you verify that it's also slow if you are using `goimports` directly?

So prepare your file so it would require an additional import to be added. E.g.: a file like this (it's missing the `import "fmt"`):

``` go
package main

func main() {
    fmt.Println("hell world")
}
```

And then call `goimports -d <path>` and see how long it takes and what it suggests to import.

Please use your code which caused `goimports` to hang/crash and not the one above.

**Comment by dominicbarnes** [comment]
Yeah, sometimes `goimports` isn't consistent about how hard it runs. I'm not saying this is something the apm package `gofmt` necessarily needs to fix, because that's probably a problem in `goimports`.

All I _really_ want is for this plugin to:
1. stop blocking my UI while that is running. this should be entirely backgrounded, and if I make another change, the running process should be killed and another one started
2. time out after a (configurable) amount of time, so it doesn't just continue unbounded (I can kill the process myself, but that's a pain)

For the time-being, I've switched to using plain `gofmt`, and I manually invoke `goimports` whenever I want to update my dep refs. I really want to use this plugin to do that magic for me, as it's really convenient, but as it stands it's more trouble than convenience. ☹️
